From 420ca8da195c6d644e61d0c4d9d6578f99c55c36 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Mon, 23 Jun 2014 21:51:30 -0400
Subject: [fs] nfsd: Helper function for parsing uuid

Message-id: <1403560314-32525-149-git-send-email-bfields@redhat.com>
Patchwork-id: 84541
O-Subject: [Patch RHEL7 148/172] NFSD: Helper function for parsing uuid
Bugzilla: 1111719
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Kinglong Mee <kinglongmee@gmail.com>

Signed-off-by: Kinglong Mee <kinglongmee@gmail.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 0d63790c365852a6ce2913632b933633343ae479
Bugzilla: 1111719
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/export.c b/fs/nfsd/export.c
index 9a41d3d..8771f41 100644
--- a/fs/nfsd/export.c
+++ b/fs/nfsd/export.c
@@ -480,6 +480,23 @@ static inline int
 secinfo_parse(char **mesg, char *buf, struct svc_export *exp) { return 0; }
 #endif
 
+static inline int
+uuid_parse(char **mesg, char *buf, unsigned char **puuid)
+{
+ int len;
+
+ /* expect a 16 byte uuid encoded as \xXXXX... */
+ len = qword_get(mesg, buf, PAGE_SIZE);
+ if (len != 16)
+  return -EINVAL;
+
+ *puuid = kmemdup(buf, 16, GFP_KERNEL);
+ if (*puuid == NULL)
+  return -ENOMEM;
+
+ return 0;
+}
+
 static int svc_export_parse(struct cache_detail *cd, char *mesg, int mlen)
 {
  /* client path expiry [flags anonuid anongid fsid] */
@@ -558,18 +575,9 @@ static int svc_export_parse(struct cache_detail *cd, char *mesg, int mlen)
   while ((len = qword_get(&mesg, buf, PAGE_SIZE)) > 0) {
    if (strcmp(buf, "fsloc") == 0)
     err = fsloc_parse(&mesg, buf, &exp.ex_fslocs);
-   else if (strcmp(buf, "uuid") == 0) {
-    /* expect a 16 byte uuid encoded as \xXXXX... */
-    len = qword_get(&mesg, buf, PAGE_SIZE);
-    if (len != 16)
-     err  = -EINVAL;
-    else {
-     exp.ex_uuid =
-      kmemdup(buf, 16, GFP_KERNEL);
-     if (exp.ex_uuid == NULL)
-      err = -ENOMEM;
-    }
-   } else if (strcmp(buf, "secinfo") == 0)
+   else if (strcmp(buf, "uuid") == 0)
+    err = uuid_parse(&mesg, buf, &exp.ex_uuid);
+   else if (strcmp(buf, "secinfo") == 0)
     err = secinfo_parse(&mesg, buf, &exp);
    else
     /* quietly ignore unknown words and anything
-- 
1.7.1